### ISSUE TITLE
Fix: Clean up and restore support for integrating DAE states - [merged]

### ISSUE THREAD

**Issue by SGeeversAtVortech** [Deltares/rtc-tools#1519]
In GitLab by @baayen on May 16, 2023, 13:39

_Merges modernise-dae-integration -> master_

The support for integrating DAE states with CollocatedIntegratedOptimizationProblem had undergone some decay.  This PR cleans up and fixes the state integration functionality.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 13:40

added 1 commit

<ul><li>bd839e7c - Remove print statements</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682258045&start_sha=633ab6c20aebf0019abe49757cf6f90ef148553b)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 17:17

added 1 commit

<ul><li>25dbe232 - Fix map_path_expression() for integrated states</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682589895&start_sha=bd839e7cb9cc9807a144302568cc4717a7ff0c50)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 17:19

added 1 commit

<ul><li>87f9fe43 - Pacify flake8</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682592865&start_sha=25dbe23278ee33ce5d4d8b9d54facb918400ef8e)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 17:25

added 1 commit

<ul><li>740f8620 - Pacify flake8</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682600610&start_sha=87f9fe431a8abc7f6564b451278f34e2500f7f7c)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 17:29

added 1 commit

<ul><li>f14c0b63 - Pacify flake8</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682606488&start_sha=740f8620a9bb232f684a1dcced9899fed5511e0c)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 20:13

added 1 commit

<ul><li>84464a9b - Avoid taking size of unset variable</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682756007&start_sha=f14c0b635d85fa463f8cb32d79fe446b0fdf5a0c)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 20:43

added 2 commits

<ul><li>50047cde - Fix implementation of map_path_expression() for integrated states</li><li>3b1c9120 - Avoid taking size of unset variable</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682779877&start_sha=84464a9b3a07565e7bb76c518dfe60a4c5376f30)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on May 16, 2023, 20:56

added 1 commit

<ul><li>c8a30832 - Set dae_residual_collocated_size = 0 in corner case of empty problem</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=682789169&start_sha=3b1c9120480b2b0e7910fcab47fe45c3cd1e5763)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Aug 17, 2023, 22:24

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 130](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cR130)



This is API breaking, so strictly speaking shouldn't be part of RTC-Tools 2.x. One can argue it's about time for a 3.x, but there is also something to be said for "no one uses integrated states, it's broken, API changes OK on this front for 2.6". 

Either way the question I have with the yes/no switch, does it still make sense to have a "CollocatedIntegratedOptimizationProblem"? This change seems to suggest it's either one or the other, but not both at the same time (as it used to be). Shouldn't we have two separate mixins? Aren't there other mixins we might want to add that are specifically aimed at extending collocation, and others specifically aimed at extending integration?

EDIT:
Would a non-API-breaking way be to specify 
```python
@property
def integrated_states(self):
    return [*self.algebraic_states, *self.differentiated_states]
```

Possibly something like

```python
@property
def integrate_states(self): -> bool
    return False

@property
def integrated_states(self): -> List[str]
    if self.integrate_states:
        return [*self.algebraic_states, *self.differentiated_states]
    else:
        return []
```

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Aug 18, 2023, 19:59

Commented on [src/rtctools/optimization/goal_programming_mixin_base.py line 497](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..29c09dfbe9a5f8b41dbe6ee111a14d5f45262111#diff-53123ece6c64465cb2a89a1b58b7bc23R497)



Is this MR still a draft, or is this a TODO for some later day?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Aug 18, 2023, 20:09

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2271](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cR2271)



Are path_variables that different from algebraic_states that they can't/shouldn't be integrated?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Aug 18, 2023, 20:09

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2196](https://github.com/Deltares/rtc-tools/compare/42cecad4733667358a47effd46647ee605c017fb..c8a308329470b8d93d8b1be0f74630b76b8249ab#diff-1fe40e6d35b6313eb1a0db0b5798809cR2196)



Is this MR still a draft, or is this a TODO for some later day?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Sep 26, 2023, 13:38

marked this merge request as **draft**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:40

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2196](https://github.com/Deltares/rtc-tools/compare/42cecad4733667358a47effd46647ee605c017fb..c8a308329470b8d93d8b1be0f74630b76b8249ab#diff-1fe40e6d35b6313eb1a0db0b5798809cR2196)



changed this line in [version 11 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=812910398&start_sha=c8a308329470b8d93d8b1be0f74630b76b8249ab#173590c4c1ff2d3d45090969738765bbe8b3c79a_2196_2697)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:40

added 57 commits

<ul><li>c8a30832...eaed7780 - 56 commits from branch <code>master</code></li><li>c52e58b3 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=812910398&start_sha=c8a308329470b8d93d8b1be0f74630b76b8249ab)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:42

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 130](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cR130)



I have implemented your second suggestion.  In addition, I've added a check that raises an exception if `integrate_state = True` but `integrated_states != [*self.algebraic_states, *self.differentiated_states]`.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:47

added 1 commit

<ul><li>77e702cb - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=812921200&start_sha=c52e58b388a6eff41f8fab2e0173b6a736b5ac1a)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:47

marked this merge request as **ready**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:48

added 1 commit

<ul><li>59b2b9d3 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=812923726&start_sha=77e702cbdfcf9463ae842f585bebf77025f64d3c)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:50

added 1 commit

<ul><li>375ce7d9 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=812926102&start_sha=59b2b9d302fed9b2c5186e5d7b642dfe6142095f)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 15:50

requested review from @vreeken

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 16:10

added 1 commit

<ul><li>b0ec7128 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=812959652&start_sha=375ce7d95b90a0220442496932d9d703e8871bf8)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 11, 2023, 16:21

added 1 commit

<ul><li>ce34a053 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=812977459&start_sha=b0ec7128f5ed19931bf5ecd24059fe01e9ada857)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 12, 2023, 08:41

added 1 commit

<ul><li>29c09dfb - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=813592964&start_sha=ce34a053528879d37a608d6f62cf4c1e2d0fa9a6)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 25, 2023, 19:46

Review very much still in progress, just keeping track of things here:

Overall comments 

Jorn:

Can you elaborate (here, and in the commit message) on:
- Why this change is needed (i.e. what exactly was broken before)
- Why we can't integrate _some_ states, but it's all or nothing. Was this always "wrong", or what changed?
- Why algebraic states are integrated, but path_variables are not

Tjerk:

- Test with vector control states
- Test with vector states/algebraic states (do we even support that for collocation?)
- Add test that, indeed if not all states integrated, we throw error
- Add test with ensembles
- Add test with custom constraints/path variables/etc
- ~~Test with path variables added, and extra constraints~~ Works fine
- ~~What happens if we set a path variable equal to an algebraic state?~~ -> works fine
- ~~Should we allow users to say whether a path_variable is integrated or not?~~ integration, theta 0 / 1, only for model (DAE) things. Not for path_variables/equations. Not saying it should stay that way, but that's what it is now (and has been since the beginning).
- ~~Does if work with qpsol? Or not possible because of rootfinder?~~ --> Indeed does not work.
 
- Why does integration use rootfinder, and collocation does not? 

Notes:
- Test for both theta 0 and 1.. do we still allow values inbetween?


General thoughts: 

- it feels kind of weird to have all this "constant_inputs_0" and "constant_inputs_1" as separate arguments to a function, mix-and-match with theta somewhere between 0 and 1, and then in the end map them to overlapping state entries (shifted by one). I guess this has a reason?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 30, 2023, 13:04

Commented on [src/rtctools/optimization/goal_programming_mixin_base.py line 497](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..29c09dfbe9a5f8b41dbe6ee111a14d5f45262111#diff-53123ece6c64465cb2a89a1b58b7bc23R497)



changed this line in [version 18 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=829795350&start_sha=29c09dfbe9a5f8b41dbe6ee111a14d5f45262111#aab4f52ba81fa07a95e1b359ba302a402e2bc6a7_497_497)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 30, 2023, 13:04

added 1 commit

<ul><li>9a6aea56 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=829795350&start_sha=29c09dfbe9a5f8b41dbe6ee111a14d5f45262111)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 30, 2023, 13:05

added 1 commit

<ul><li>b51ffe46 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=829796474&start_sha=9a6aea56290773949682c9dc0cd9a0596dd3010d)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 30, 2023, 13:10

added 1 commit

<ul><li>071018f6 - Modernise DAE integration functionality: either integrate all states, or none at all</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=829803927&start_sha=b51ffe461f1d541ff29468d14a85ec7a076f6f6b)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 30, 2023, 13:22

1) Before, there were no tests combining integrated states with goal programming.  Several pieces of functionality, such as path variables, goals, and so forth, were not being tested in combination with integrated states.  As a result, changes were made to the goal programming implementation (as well as to the CollInt core) that broke these combinations.  This PR introduces, besides the fixes, new tests for the previously broken combinations.

2) Integrating "some" states works when the number of model equations that depends on the integrated states is equal to the number of integrated states.  Such an equality would enable one to decompose the model into integrated and non-integrated parts.  In general, however, there will be cross-dependencies and for this reason the equality will not hold.  This PR simplifies the logic such that either all states are integrated, or none.

3) Integration is only applicable to model states.  Path variables are auxiliary variables at the level of the optimization problem, i.e., are introduced at a different level.  The model sits at the core of RTC-Tools; around the core, it offers functionalities such as goal programming which wrap around, but do not modify, the model.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 30, 2023, 13:23

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2271](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cR2271)



See discussion below

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 30, 2023, 17:48

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 911](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..4dee712e88abfb0f92f61f0331b91210fe589cc3#diff-1fe40e6d35b6313eb1a0db0b5798809cR911)



I'd say this is more like an "elif" now, not a separate if

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 30, 2023, 21:00

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1081](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..1503fa951520e5ee96a801a7e94fd5019d3e9fa1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1081)



Same, more like an `elif`. But this one is a bit different as there is already an `else` above (that may or may not be needed anymore)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:25

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 911](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..4dee712e88abfb0f92f61f0331b91210fe589cc3#diff-1fe40e6d35b6313eb1a0db0b5798809cR911)



changed this line in [version 22 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=830967208&start_sha=4dee712e88abfb0f92f61f0331b91210fe589cc3#173590c4c1ff2d3d45090969738765bbe8b3c79a_911_911)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:25

added 1 commit

<ul><li>1503fa95 - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=830967208&start_sha=4dee712e88abfb0f92f61f0331b91210fe589cc3)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:36

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1081](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..1503fa951520e5ee96a801a7e94fd5019d3e9fa1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1081)



changed this line in [version 23 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=830982819&start_sha=1503fa951520e5ee96a801a7e94fd5019d3e9fa1#173590c4c1ff2d3d45090969738765bbe8b3c79a_1081_1077)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:36

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1094](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..1503fa951520e5ee96a801a7e94fd5019d3e9fa1#diff-1fe40e6d35b6313eb1a0db0b5798809cL1094)



changed this line in [version 23 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=830982819&start_sha=1503fa951520e5ee96a801a7e94fd5019d3e9fa1#173590c4c1ff2d3d45090969738765bbe8b3c79a_1090_1085)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:36

added 1 commit

<ul><li>0dcadcaf - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=830982819&start_sha=1503fa951520e5ee96a801a7e94fd5019d3e9fa1)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:37

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 911](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..4dee712e88abfb0f92f61f0331b91210fe589cc3#diff-1fe40e6d35b6313eb1a0db0b5798809cR911)



Good catch!  I've pushed a fix.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:37

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1081](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..1503fa951520e5ee96a801a7e94fd5019d3e9fa1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1081)



Good catch!  I've pushed a fix.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 10:38

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1094](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..1503fa951520e5ee96a801a7e94fd5019d3e9fa1#diff-1fe40e6d35b6313eb1a0db0b5798809cL1094)



Good catch!  This is no longer needed now that we either integrate all states, or none at all.  I've pushed a fix.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 11:33

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1354](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..0dcadcafbd98ce4add3b900c1346e0a899724904#diff-1fe40e6d35b6313eb1a0db0b5798809cR1354)



Am I correct that a different order would break things? If so, it's probbaly better to indeed rely on an order defined _once_ (like it was before). There is `integrated_variable_names` available with the correct order afaict

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 11:46

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1354](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..0dcadcafbd98ce4add3b900c1346e0a899724904#diff-1fe40e6d35b6313eb1a0db0b5798809cR1354)



changed this line in [version 24 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831087705&start_sha=0dcadcafbd98ce4add3b900c1346e0a899724904#173590c4c1ff2d3d45090969738765bbe8b3c79a_1354_1354)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 11:46

added 1 commit

<ul><li>ee269abc - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831087705&start_sha=0dcadcafbd98ce4add3b900c1346e0a899724904)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 11:46

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1354](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..0dcadcafbd98ce4add3b900c1346e0a899724904#diff-1fe40e6d35b6313eb1a0db0b5798809cR1354)



Good catch.  Pushed a change for this too.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:15

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1561](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1561)



You can use `integrated_variable_names` here instead, saves you the `.name()` lookup

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:18

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1580](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1580)



Again a bit dangerous to rely on this order. Use `integated_variable_names` instead.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [tests/optimization/data/HybridShootingModel.mo line 1](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-977337a1a12985db2e1f644c7c713aa5L1)



Might as well rename file and the model to SingleShooting for consistency

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [tests/optimization/data/HybridShootingModel.mo line 4](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-977337a1a12985db2e1f644c7c713aa5R4)



Why did you add this?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 553](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cL553)



This only says something about collocation. Should it be fixed to also mention integration? Or should we remove the warning altogether if we intend to keep support theta < 1

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 900](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cL900)



Might as well delete this commented out code while you're fixing/cleaning up

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commit message: please wrap paragraphs (if using vim `shift+V -> g q`)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1622](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cR1622)



To save the cost of the `.name()`, there is `dae_constant_inputs_names` available

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1629](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cR1629)



Likewise, `self.extra_constant_inputs_name_and_size` (but yeh, comes with size)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1995](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cL1995)



This change is not necessary right? You already have the check that `if self.integrated_state, then self.integrated_states == [...]` elsewhere

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2127](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cR2127)



Likewise

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [setup.py line 53](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-2eeaed663bd0d25b7e608891384b7298R53)



Good catch, but separate commit please

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2354](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cR2354)



I guess same comment, you can leave it `self.integrated_states`, but if other changes nearby are needed you can leave it

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1572](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cL1572)



This _is_ an API change I guess. Might be worth mentioning in the commit message with the other API changes

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2356](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cR2356)



When is it a slice? I thought it's always a singular value when integrating. In other words, I don't quite understand why this change block is needed

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 12:49

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2463](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cR2463)



Hmm, was it broken for ensembles?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1561](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1561)



changed this line in [version 25 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831191051&start_sha=ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#173590c4c1ff2d3d45090969738765bbe8b3c79a_1561_1560)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1580](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1580)



changed this line in [version 25 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831191051&start_sha=ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#173590c4c1ff2d3d45090969738765bbe8b3c79a_1580_1577)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

Commented on [tests/optimization/data/HybridShootingModel.mo line 1](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-977337a1a12985db2e1f644c7c713aa5L1)



changed this line in [version 25 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831191051&start_sha=ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#7ed16c4a648b57ff187e42cf22abfd8585c0db56_1_0)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

Commented on [tests/optimization/data/HybridShootingModel.mo line 4](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-977337a1a12985db2e1f644c7c713aa5R4)



changed this line in [version 25 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831191051&start_sha=ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#7ed16c4a648b57ff187e42cf22abfd8585c0db56_4_0)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

added 1 commit

<ul><li>603e6cce - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831191051&start_sha=ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1561](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1561)



Yup, good catch

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1580](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-1fe40e6d35b6313eb1a0db0b5798809cR1580)



Indeed

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:56

Commented on [tests/optimization/data/HybridShootingModel.mo line 1](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-977337a1a12985db2e1f644c7c713aa5L1)



Agreed

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:57

Commented on [tests/optimization/data/HybridShootingModel.mo line 4](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..ee269abc8fe42e5cec56cd2ebeab9f22bcf80dd1#diff-977337a1a12985db2e1f644c7c713aa5R4)



Integration wasn't being tested in combination with algebraic variables.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:58

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2463](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cR2463)



Yup, that too

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 12:59

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1572](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-1fe40e6d35b6313eb1a0db0b5798809cL1572)



I don't think `integrators` was ever meant to be exposed

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:17

Commented on [setup.py line 53](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-2eeaed663bd0d25b7e608891384b7298R53)



Will do this later, just ahead of merge, once the main commit is complete to avoid having to perform too much git surgery

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:19

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 553](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cL553)



Adding a mention of integration

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:23

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 553](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cL553)



changed this line in [version 26 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831229515&start_sha=603e6cce9914c6d062750df48d28c2453df0c83e#173590c4c1ff2d3d45090969738765bbe8b3c79a_556_556)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:23

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 900](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cL900)



changed this line in [version 26 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831229515&start_sha=603e6cce9914c6d062750df48d28c2453df0c83e#173590c4c1ff2d3d45090969738765bbe8b3c79a_892_892)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:23

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1622](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cR1622)



changed this line in [version 26 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831229515&start_sha=603e6cce9914c6d062750df48d28c2453df0c83e#173590c4c1ff2d3d45090969738765bbe8b3c79a_1622_1621)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:23

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1629](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..603e6cce9914c6d062750df48d28c2453df0c83e#diff-1fe40e6d35b6313eb1a0db0b5798809cR1629)



changed this line in [version 26 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831229515&start_sha=603e6cce9914c6d062750df48d28c2453df0c83e#173590c4c1ff2d3d45090969738765bbe8b3c79a_1629_1627)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:23

added 1 commit

<ul><li>9b7cefb4 - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831229515&start_sha=603e6cce9914c6d062750df48d28c2453df0c83e)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:26

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2356](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cR2356)



Good catch!

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:33

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 1995](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cL1995)



changed this line in [version 27 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831243012&start_sha=9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#173590c4c1ff2d3d45090969738765bbe8b3c79a_2048_2044)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:33

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2127](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cR2127)



changed this line in [version 27 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831243012&start_sha=9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#173590c4c1ff2d3d45090969738765bbe8b3c79a_2127_2124)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:33

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2354](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cR2354)



changed this line in [version 27 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831243012&start_sha=9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#173590c4c1ff2d3d45090969738765bbe8b3c79a_2354_2348)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:33

Commented on [src/rtctools/optimization/collocated_integrated_optimization_problem.py line 2356](https://github.com/Deltares/rtc-tools/compare/eaed7780b4102ef926e33ce550edb8e93d1a38fd..9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#diff-1fe40e6d35b6313eb1a0db0b5798809cR2356)



changed this line in [version 27 of the diff](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831243012&start_sha=9b7cefb4826c11aeff2a4ce76fbee3217d6e6465#173590c4c1ff2d3d45090969738765bbe8b3c79a_2356_2350)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:33

added 1 commit

<ul><li>635e8ebb - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831243012&start_sha=9b7cefb4826c11aeff2a4ce76fbee3217d6e6465)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:35

added 1 commit

<ul><li>75112666 - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831246509&start_sha=635e8ebb7eeeabb6d9f832edd79ab62e6b5d46d1)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 13:42

added 1 commit

<ul><li>12500499 - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=831255987&start_sha=751126666e722486284f15cf89e24a41679daaa2)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Oct 31, 2023, 14:58

Commented on [tests/optimization/test_integration.py line 128](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-aeca87bc8a3dac4a5945faaf4474c5c7R128)



I have to say I find it a little confusing to have Goal1 not be the first goal to solve for.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Oct 31, 2023, 15:05

Commented on [tests/optimization/test_integration.py line 128](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-aeca87bc8a3dac4a5945faaf4474c5c7R128)



Agreed - it's a copy/paste from test_goal_programming.py though ..

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 1, 2023, 18:50

Commented on [tests/optimization/test_integration.py line 128](https://github.com/Deltares/rtc-tools/compare/a3bd87c4de855668cef43877db6c679ac93a9db7..fa26ce022a4ca678502ebee486a2746999397a72#diff-aeca87bc8a3dac4a5945faaf4474c5c7R128)



Then keep it as is. Consistency above all I guess, even if it's misleading 😅

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 1, 2023, 18:58

I've been playing around with making some additional tests on the `tests-modernize-dae-integration` branch. One thing I run into, and that I'm not sure how to deal with yet, is the `integrate_states` and `integrated_states` both existing, and doing cross-checks (in the tests, so users don't cut themselves).

If I'm honest, the solution that I'd prefer from a backwards compatibility standpoint is:
- No `self.integrate_states`, only `self.integrated_states`. And then a check that it covers all states, and not just some of them (as it is now). With this option, if someone has a model that did the following, it will just keep on working with the changes of this MR:

```python 
def integrated_states(self):
    return [*self.algebraic_states, *self.differentiated_states]
```

From a ease-of-use standpoint, e.g. for all models made with integration in the future, I prefer the opposite:
- No `self.integrated_states`, only `self.integrate_states` (True/False). And then, because we're breaking the API, check that calling `self.integrated_states` results in an AttributeError.

Thoughts @baayen ? I realize I'm hypothetically asking you to undo what I requested before 😅

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 2, 2023, 11:06

Would you like to add a commit to this PR with your preferred implementation of the setting? :smile:

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 2, 2023, 12:00

added 1 commit

<ul><li>5d8d37be - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=833472346&start_sha=12500499174b8a7412803aaef355b48cf38e942a)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 2, 2023, 12:08

added 1 commit

<ul><li>f141d351 - CollIntOptimizationProblem:  Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=833483851&start_sha=5d8d37be3d4c2931c5678a4cbf85da26796e430b)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 6, 2023, 13:37

added 2 commits

<ul><li>c4b9cfc8 - setup.py: Fix missing test dependency on NetCDF4</li><li>3356d2cb - CollIntOptimizationProblem: Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=836186690&start_sha=f141d351230b5c2005fbda2f0905ff1b08abcad7)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 6, 2023, 13:55

added 11 commits

<ul><li>3356d2cb...521f51f6 - 8 commits from branch <code>master</code></li><li>de45f44e - setup.py: Fix missing test dependency on NetCDF4</li><li>cc90c117 - CollIntOptimizationProblem: Fix and modernise DAE integration functionality</li><li>9f9829bb - API changes and test</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=836208902&start_sha=3356d2cb991f3fe4db3e582ae2289695d4aa8d6b)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 6, 2023, 14:03

After a lot of back and forth, I chose ease-of-use for future use. I really _really_ hold API backwards compability in high regard, but because we can safely say no one was using the old  (because it was miserably broken), it feels weird to burden future use with a quirky API. Everyone would always have to do:

```python 
# Ugly, because it cannot be anything other than this combination of states
def integrated_states(self):
    return [*self.algebraic_states, *self.differentiated_states]
```

instead of 

```python
# Much cleaner
def integrated_states(self):
    return True
```

Also added a check and a test that anyone accidentally using the old API is steered in the right direction. I very much doubt anyone would run into this, but do want to make sure we have a backstop.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 6, 2023, 14:03

resolved all threads

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 6, 2023, 14:07

@baayen Can you review (and squash if OK) my last commit into your big one?

Note:
- I rebased on top of master
- Split out the NetCDF test dependency fix into a separate commit

I force-pushed because I know you don't particularly like the git surgery. If for some reason you had local changes that would conflict with this force-pushing, I made a backup prior that we can revert to (`backup-modernise-dae-integration-202311061330`)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 11:03

added 1 commit

<ul><li>0e397cdc - CollIntOptimizationProblem: Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=837281380&start_sha=9f9829bbfdc6b7360a0fe140fd6899a93352813f)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 11:04

I'm good with this approach. I've squashed the commits.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 11:14

added 1 commit

<ul><li>edf7445b - CollIntOptimizationProblem: Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=837297296&start_sha=0e397cdc31505655d46684ad04032f339612477c)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 11:15

I have added a minor change removing the `__integrated_states` property, instead setting is as an instance variable in `transcribe()`.  This is consistent with the use of other private variables such as `__integrators`.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 11:23

added 1 commit

<ul><li>8fb51ee5 - CollIntOptimizationProblem: Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=837312682&start_sha=edf7445bd271cfc3221ecf2d84724072994df77b)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 7, 2023, 13:07

approved this merge request

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 7, 2023, 16:48

added 1 commit

<ul><li>576c164e - optimization: fix small type error</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=837794422&start_sha=8fb51ee51f3b8bd19670250103e5b6db133e65f0)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 7, 2023, 16:53

I'm not sure if I fully understand the difference between collocation/integration. Could you expand the docstring of CollocatedIntegratedOptimizationProblem to explain this difference a bit more?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 17:39

added 1 commit

<ul><li>c2f54172 - CollIntOptimizationProblem: Fix and modernise DAE integration functionality</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=837875928&start_sha=576c164e6c6acc32f1543311bc106dbe9f3ea719)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 17:40

Sure. I have just added another paragraph to the docstring

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 7, 2023, 17:40

resolved all threads

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Nov 13, 2023, 08:33

added 1 commit

<ul><li>17fef32b - test_modelica_mixin:  Test collocated variables against their discretization</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=842397681&start_sha=c2f54172f23d8acf8191dabbd977c933e0debbc4)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 14, 2023, 10:45

I still don't fully understand the difference between collocation and integration and what the advantage is of one over the other. I see that a rootfinder is used for integration similar to the implicit time stepping in simulation. Is this the main difference: that the constraints of the optimization problem contain a rootfinder when using integration? And is this more efficient than simply using the discretized equations as constraints?

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 14, 2023, 10:45

Apart from this, I didn't see anything suspicious in this branch, although I understand very little.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Nov 14, 2023, 12:14

Definitely might be interesting to add that explanation, for example when collocation or integration is most suited to solve a problem. And I also don't fully understand it all either, like how (if at all) are variable bounds satisfied by nlpsol when there's only the single state at the start. And if CasADi does anything smart with the (recursive?) rootfinding with the mapaccum.

Either way, I don't think it's really in scope of _this_ MR to explain all that. This MR just restores/fixes integration, and adds more tests to make sure we don't easily break it again.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 14, 2023, 14:00

Fair enough since this is a fix of something that was already there. I've made an issue that we should add this documentation. It is quite useful if we understand the difference and know when to use one or the other method.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 14, 2023, 14:00

resolved all threads

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 14, 2023, 14:00

approved this merge request

**Comment by SGeeversAtVortech** [comment]
In GitLab by @SGeeversAtVortech on Nov 14, 2023, 14:01

added 11 commits

<ul><li>17fef32b...a3bd87c4 - 8 commits from branch <code>master</code></li><li>58efa3bd - setup.py: Fix missing test dependency on NetCDF4</li><li>a6a34c8c - CollIntOptimizationProblem: Fix and modernise DAE integration functionality</li><li>fa26ce02 - test_modelica_mixin:  Test collocated variables against their discretization</li></ul>

[Compare with previous version](/deltares/rtc-tools/-/merge_requests/440/diffs?diff_id=843969668&start_sha=17fef32bda0d6cc29f2b11e336b2b98b420a9ac3)
